File: fe-ta/src/models/Product.ts
import type { ProductColorResponse } from "./ProductColor";
import type { ProductMediaResponse } from "./ProductMedia";
import type { ProductSpecResponse } from "./ProductSpec";
import type { ProductStorageResponse } from "./ProductStorage";

export interface ProductRequest {
  categoryId?: number;
  productName: string;
  slug: string;
  shortDescription?: string;
  longDescription?: string;
  originalPrice?: number;
  discountPrice?: number;
  isFeatured?: boolean;
  isActive?: boolean;
  viewsCount?: number;
  note?: string;

}

export interface ProductResponse {
  id: number;
  categoryId: number;
  categoryName?: string;
  productName: string;
  slug: string;
  shortDescription?: string;
  longDescription?: string;
  originalPrice?: number;
  discountPrice?: number;
  discount?: number;
  isFeatured: boolean;
  isActive: boolean;
  viewsCount: number;
  createUid?: number;
  writeIUid?: number;
  updateTime?: string; 
  createdName?: string;
  updatedName?: string;
  note?: string;
  option1?: string;
  option2?: string;
  option3?: string;
  option4?: string;
  option5?: string;

  mediaList?: ProductMediaResponse[];
  specs?: ProductSpecResponse[];
  colors?: ProductColorResponse[];
  storages?: ProductStorageResponse[];
}

export interface ProductFilterRequest {
  productName?: string;
  categoryName?: string;
  slug?: string;
  shortDescription?: string;
  longDescription?: string;

  // Lọc theo khoảng giá
  minPrice?: number;
  maxPrice?: number;

  isFeatured?: boolean;
  isActive?: boolean;

  // Lọc theo views
  minViewsCount?: number;
  maxViewsCount?: number;

  // User tạo/sửa
  createdName?: string;
  updatedName?: string;

  // Lọc theo thời gian
  fromUpdateTime?: string; // ISO string (yyyy-MM-dd)
  toUpdateTime?: string;

  note?: string;
}

File: beSQLSugar/beSQLSugar/Infrastructure/Database/Enities/Product.cs
using SqlSugar;
namespace beSQLSugar.Infrastructure.Database.Enities
{
    [SugarTable("products")]
    public class Product
    {
        [SugarColumn(IsPrimaryKey = true, IsIdentity = true)]
        public int Id { get; set; }

        [SugarColumn(IsNullable = false)]
        public int CategoryId { get; set; } // liên kết với Category

        [SugarColumn(ColumnDataType = "nvarchar(250)", IsNullable = false)]
        public string? ProductName { get; set; }

        [SugarColumn(ColumnDataType = "nvarchar(250)", IsNullable = false)]
        public string? Slug { get; set; } 

        [SugarColumn(ColumnDataType = "nvarchar(250)")]
        public string? ShortDescription { get; set; }

        [SugarColumn(ColumnDataType = "nvarchar(max)")]
        public string? LongDescription { get; set; }

        [SugarColumn(IsNullable = true)]
        public decimal OriginalPrice { get; set; }
        [SugarColumn(IsNullable = true)]
        public decimal DiscountPrice { get; set; }
        [SugarColumn(IsNullable = true)]
        public int Discount { get; set; }
        [SugarColumn(DefaultValue = "0")]
        public bool IsFeatured { get; set; } = false;

        [SugarColumn(DefaultValue = "1")]
        public bool IsActive { get; set; } = true;

        [SugarColumn(DefaultValue = "0")]
        public int ViewsCount { get; set; } = 0;
        [SugarColumn(IsNullable = true)]
        public int? CreateUid { get; set; }

        [SugarColumn(ColumnDataType = "NVARCHAR(100)", IsNullable = true)]
        public string? CreatedName { get; set; }

        [SugarColumn(IsNullable = true)]
        public int? WriteIUid { get; set; }

        [SugarColumn(ColumnDataType = "NVARCHAR(100)", IsNullable = true)]
        public string? UpdatedName { get; set; }

        [SugarColumn(IsNullable = true)]
        public DateTime? UpdateTime { get; set; }

        [SugarColumn(ColumnDataType = "nvarchar(250)", IsNullable = true)]
        public string? Note { get; set; }

        [SugarColumn(ColumnDataType = "nvarchar(120)", IsNullable = true)]
        public string? Option1 { get; set; }

        [SugarColumn(ColumnDataType = "nvarchar(120)", IsNullable = true)]
        public string? Option2 { get; set; }

        [SugarColumn(ColumnDataType = "nvarchar(120)", IsNullable = true)]
        public string? Option3 { get; set; }

        [SugarColumn(ColumnDataType = "nvarchar(120)", IsNullable = true)]
        public string? Option4 { get; set; }

        [SugarColumn(ColumnDataType = "nvarchar(120)", IsNullable = true)]
        public string? Option5 { get; set; }

        // 🔑 Navigation: 1 Product -> nhiều Media
        [Navigate(NavigateType.OneToMany, nameof(ProductMedia.ProductId))]
        public List<ProductMedia>? MediaList { get; set; }

        // 🔑 Navigation: 1 Product -> nhiều Specs
        [Navigate(NavigateType.OneToMany, nameof(ProductSpec.ProductId))]
        public List<ProductSpec>? Specs { get; set; }

        [Navigate(NavigateType.OneToOne, nameof(CategoryId))]
        public Category? Category { get; set; }

        [Navigate(NavigateType.OneToMany, nameof(ProductColor.ProductId))]
        public List<ProductColor>? Colors { get; set; }

        [Navigate(NavigateType.OneToMany, nameof(ProductStorage.ProductId))]
        public List<ProductStorage>? Storages { get; set; }
    }
}
